utilities/pspp-convert.1
.\" -*- nroff -*-
.de IQ
.  br
.  ns
.  IP "\\$1"
..
.TH pspp\-convert 1 "October 2013" "PSPP" "PSPP Manual"
.
.SH NAME
pspp\-convert \- convert SPSS system and portable files to other formats
.
.SH SYNOPSIS
\fBpspp\-convert\fR [\fIoptions\fR] \fIinput\fR \fIoutput\fR
.br
\fBpspp\-convert \-\-help\fR | \fB\-h\fR
.br
\fBpspp\-convert \-\-version\fR | \fB\-v\fR
.
.SH DESCRIPTION
The \fBpspp\-convert\fR program reads \fIinput\fR, which may be an
SPSS system file, an SPSS/PC+ system file, or an SPSS portable file,
and writes it to \fIoutput\fR, performing format conversion as
necessary.
.PP
The format of \fIinput\fR is automatically detected, when possible.
The character encoding of old SPSS system files cannot always be
guessed correctly, and SPSS/PC+ system files do not include any
indication of their encoding.  Use \fB\-e \fIencoding\fR to specify
the encoding in this case.
.PP
By default, the intended format for \fIoutput\fR is inferred from its
extension:
.
.IP \fBcsv\fR
.IQ \fBtxt\fR
Comma-separated value.  Each value is formatted according to its
variable's print format.  The first line in the file contains variable
names.
.
.IP \fBsav\fR
.IQ \fBsys\fR
SPSS system file.
.
.IP \fBpor\fR
SPSS portable file.
.
.PP
Use \fB\-O \fIextension\fR to override the inferred format or to
specify the format for unrecognized extensions.
.
.PP
\fBpspp\-convert\fR can convert most input formats to most output
formats, with one exception: if the input file is an encrypted system
file, then the output file must also be an (unencrypted) system file.
.
.SH "OPTIONS"
.
.IP "\fB\-O format\fR"
.IQ "\fB\-\-output\-format=\fIformat\fR"
Specifies the desired output format.  \fIformat\fR must be one of the
extensions listed above, e.g. \fB\-O csv\fR requests comma-separated
value output.
.
.IP "\fB\-c \fImaxcases\fR"
.IQ "\fB\-\-cases=\fImaxcases\fR"
By default, all cases are copied from \fIinput\fR to \fIoutput\fR.
Specifying this option to limit the number of cases written to
\fIoutput\fR to \fImaxcases\fR.
.
.IP "\fB\-e \fIcharset\fR"
.IQ "\fB\-\-encoding=\fIcharset\fR"
Overrides the encoding in which character strings in \fIinput\fR are
interpreted.  This option is necessary because old SPSS system files
do not self-identify their encoding.
.
.IP "\fB\-p \fIpassword\fR"
.IQ "\fB\-\-password=\fIpassword\fR"
Specifies the password to use to decrypt an encrypted SPSS system file
\fIinput\fR.  If this option is not specified, \fBpspp\-convert\fR
prompts for the password.
.
.IP
On multiuser systems, this option may not be safe because other users
may be able to see the password in process listings.
.
.IP "\fB\-h\fR"
.IQ "\fB\-\-help\fR"
Prints a usage message on stdout and exits.
.
.IP "\fB\-v\fR"
.IQ "\fB\-\-version\fR"
Prints version information on stdout and exits.
.
.SH "AUTHORS"
Ben Pfaff.
.
.SH "SEE ALSO"
.
.BR pspp (1),
.BR psppire (1).


NEWS
PSPP NEWS -- history of user-visible changes.
Copyright (C) 1996-2000, 2008-2014 Free Software Foundation, Inc.
See the end for copying conditions.

Please send PSPP bug reports to bug-gnu-pspp@gnu.org.
 
Changes since 0.8.4:

 * SPSS/PC+ system files are now supported on GET and other commands
   that read SPSS system files.  The pspp-convert program can now read
   SPSS/PC+ system files.  Writing the obsolete SPSS/PC+ system file
   format is not supported.

 * SYSFILE INFO can now read SPSS/PC+ system files and SPSS portable
   files.

 * FREQUENCIES: A bug was fixed where an assertion failure occured
   when an empty dataset was presented.

 * The GRAPH command is now available.  Initially it supports
   scatterplots and histograms.

Changes from 0.8.3 to 0.8.4:

 * Formatting of SYSFILE INFO output was made easier to read.

 * Bug fixes, including the following notable ones:

   - FREQUENCIES works properly for string variables.  (This bug was
     introduced in 0.8.2.)

   - CROSSTABS now correctly computes all of the measures that it
     offers.  Some measures have been removed because they were not
     computed correctly.

   - The NPAR TESTS calculation of significance for the RUNS
     subcommand has been corrected.

   - Planned comparisons in ONEWAY ANOVA now correctly handle negative
     T-values.

   - Conformance fixes to Open Document output format.

Changes from 0.8.2 to 0.8.3:

 * Independent T-Tests displayed the standard error of the difference
   and the confidence interval for the unequal variances case, where
   the equal variances case should have been.  This has been fixed.

 * REGRESSION now recognises /STATISTICS=CI(x) which causes confidence
   intervals for the coefficients to be printed.

 * When DESCRIPTIVES calculates Z scores, it now omits cases filtered
   by the current FILTER settings.

 * PSPPIRE graphical user interface improvements:

   - Dialog boxes with source  variable lists can now choose the sort
     order;  right click and select from the popup menu.

   - File|Open now allows an encoding to be selected for opening
     system files.

   - File|Display Data File Information|External File... now allows an
     encoding to be selected.

   - A problem with the Means dialog has been resolved (bug #41433).

   - Several problems related to the input of spreadsheets have been
     fixed (bug #41657, #41620

 * System file related improvements:

   - With ENCODING="DETECT", SYSFILE INFO can now help the user to
     pick an encoding for reading a system file that does not identify
     its own encoding

   - SYSFILE INFO now accepts an ENCODING subcommand to specify the
     character encoding of string data in the system file.

   - Variable labels over 255 bytes long are now accepted without
     truncation (bug #41863).

   - System files that contain duplicate variable names may now be
     read successfully (bug #41475).

Changes from 0.8.1 to 0.8.2:

 * Charts are now rendered with colours from the Tango palette instead
   of fully saturated primaries.

 * Support for new system file variants:

   - PSPP can now read and write ZCOMPRESSED system files, which
     compress data much more effectively than older "compressed"
     files.  (The older format is still supported.)

   - PSPP can now decrypt encrypted system files, using the new
     pspp-convert utility.  The encrypted system file format is
     unacceptably insecure, so to discourage its use PSPP and PSPPIRE
     do not directly read or write this format.

 * Missing values for long string variables are now read from and
   written to system files in an SPSS-compatible fashion.

   (Earlier versions of PSPP that supported missing values for long
   string variables wrote them to system files in an SPSS-incompatible
   way.  To fix the problem, read the system file with this version of
   PSPP and then save a new copy of it.)

 * pspp-convert, a new standalone utility for converting SPSS system
   and portable files to other formats, is now included.  The initial
   version supports comma-separated value files as output format.
   pspp-convert can also decrypt encrypted system files.

 * Build changes:

   - zlib is now a required dependency.  (Previously it was optional.)

Changes from 0.8.0 to 0.8.1:

 * New commands:

   - VARIABLE ROLE.

 * Changes to existing commands:

   - CROSSTABS: The requirement that one expected count must be less than five
	before Fishers Exact Test will be reported has been removed.

 * PSPPIRE graphical user interface improvements:

   - A new set of icons has been contributed.

   - Syntax windows now parse syntax in "auto" mode, which in practice
     should mean that both "batch" and "interactive" syntax now works,
     instead of just "interactive" syntax.

   - The variable pane of data windows now include a Role column.

 * Notable bug fixes:

   - System files written by IBM SPSS 21 are now read without warnings.

   - System files written by "VOXCO INTERVIEWER 4.3" are now read
     without warnings.

   - PSPPIRE should now more gracefully handle syntax files that contain
     errors.

Changes from 0.6.2 to 0.8.0:

 * New commands:

   - ADD FILES
   - CORRELATIONS
   - DATAFILE ATTRIBUTES
   - DATASET ACTIVATE
   - DATASET CLOSE
   - DATASET COPY
   - DATASET DECLARE
   - DATASET DISPLAY
   - DATASET NAME
   - LOGISTIC REGRESSION
   - MATCH FILES
   - MEANS
   - MRSETS
   - PRESERVE and RESTORE
   - QUICK CLUSTER
   - RELIABILITY
   - ROC
   - SAVE TRANSLATE to CSV and tab-delimited files
   - UPDATE
   - VARIABLE ATTRIBUTES

 * Changes to existing commands:

   - AUTORECODE has a new GROUP subcommand.

   - CROSSTABS has been re-implemented to fix numerous bugs.

   - EXAMINE: /MISSING=LISTWISE is now the default.

   - DO REPEAT command has been reimplemented.  Now, when DO REPEAT
     contains an INCLUDE or INSERT command, substitutions are not
     applied to the included file.
  

utilities/pspp-convert.c
/* PSPP - a program for statistical analysis.
   Copyright (C) 2013, 2014 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>. */

#include <config.h>

#include <errno.h>
#include <getopt.h>
#include <limits.h>
#include <stdlib.h>
#include <unistd.h>

#include "data/any-reader.h"
#include "data/casereader.h"
#include "data/casewriter.h"
#include "data/csv-file-writer.h"
#include "data/file-name.h"
#include "data/por-file-writer.h"
#include "data/settings.h"
#include "data/sys-file-encryption.h"
#include "data/sys-file-writer.h"
#include "data/file-handle-def.h"
#include "libpspp/assertion.h"
#include "libpspp/cast.h"
#include "libpspp/i18n.h"

#include "gl/error.h"
#include "gl/getpass.h"
#include "gl/progname.h"
#include "gl/version-etc.h"

#include "gettext.h"
#define _(msgid) gettext (msgid)

static void usage (void);

static void decrypt_sav_file (struct encrypted_sys_file *enc,
                              const char *input_filename,
                              const char *output_filename,
                              const char *password);

int
main (int argc, char *argv[])
{
  const char *input_filename;
  const char *output_filename;

  long long int max_cases = LLONG_MAX;
  struct dictionary *dict;
  struct casereader *reader;
  struct file_handle *input_fh;
  const char *encoding = NULL;
  struct encrypted_sys_file *enc;

  const char *output_format = NULL;
  struct file_handle *output_fh;
  struct casewriter *writer;
  const char *password = NULL;

  long long int i;

  set_program_name (argv[0]);
  i18n_init ();
  fh_init ();
  settings_init ();

  for (;;)
    {
      static const struct option long_options[] =
        {
          { "cases",    required_argument, NULL, 'c' },
          { "encoding", required_argument, NULL, 'e' },
          { "password", required_argument, NULL, 'p' },

          { "output-format", required_argument, NULL, 'O' },

          { "help",    no_argument,       NULL, 'h' },
          { "version", no_argument,       NULL, 'v' },
          { NULL,      0,                 NULL, 0 },
        };

      int c;

      c = getopt_long (argc, argv, "c:e:p:O:hv", long_options, NULL);
      if (c == -1)
        break;

      switch (c)
        {
        case 'c':
          max_cases = strtoull (optarg, NULL, 0);
          break;

        case 'e':
          encoding = optarg;
          break;

        case 'p':
          password = optarg;
          break;

        case 'O':
          output_format = optarg;
          break;

        case 'v':
          version_etc (stdout, "pspp-convert", PACKAGE_NAME, PACKAGE_VERSION,
                       "Ben Pfaff", "John Darrington", NULL_SENTINEL);
          exit (EXIT_SUCCESS);

        case 'h':
          usage ();
          exit (EXIT_SUCCESS);

        default:
          exit (EXIT_FAILURE);
        }
    }

  if (optind + 2 != argc)
    error (1, 0, _("exactly two non-option arguments are required; "
                   "use --help for help"));

  input_filename = argv[optind];
  output_filename = argv[optind + 1];
  if (output_format == NULL)
    {
      const char *dot = strrchr (output_filename, '.');
      if (dot == NULL)
        error (1, 0, _("%s: cannot guess output format (use -O option)"),
               output_filename);

      output_format = dot + 1;
    }

  if (encrypted_sys_file_open (&enc, input_filename) > 0)
    {
      if (strcmp (output_format, "sav") && strcmp (output_format, "sys"))
        error (1, 0, _("can only convert encrypted data file to sav or sys "
                       "format"));

      decrypt_sav_file (enc, input_filename, output_filename, password);
      goto exit;
    }

  input_fh = fh_create_file (NULL, input_filename, fh_default_properties ());
  reader = any_reader_open_and_decode (input_fh, encoding, &dict, NULL);
  if (reader == NULL)
    exit (1);

  output_fh = fh_create_file (NULL, output_filename, fh_default_properties ());
  if (!strcmp (output_format, "csv") || !strcmp (output_format, "txt"))
    {
      struct csv_writer_options options;

      csv_writer_options_init (&options);
      options.include_var_names = true;
      writer = csv_writer_open (output_fh, dict, &options);
    }
  else if (!strcmp (output_format, "sav") || !strcmp (output_format, "sys"))
    {
      struct sfm_write_options options;

      options = sfm_writer_default_options ();
      writer = sfm_open_writer (output_fh, dict, options);
    }
  else if (!strcmp (output_format, "por"))
    {
      struct pfm_write_options options;

      options = pfm_writer_default_options ();
      writer = pfm_open_writer (output_fh, dict, options);
    }
  else
    {
      error (1, 0, _("%s: unknown output format (use -O option)"),
             output_filename);
      NOT_REACHED ();
    }

  for (i = 0; i < max_cases; i++)
    {
      struct ccase *c;

      c = casereader_read (reader);
      if (c == NULL)
        break;

      casewriter_write (writer, c);
    }

  if (!casereader_destroy (reader))
    error (1, 0, _("%s: error reading input file"), input_filename);
  if (!casewriter_destroy (writer))
    error (1, 0, _("%s: error writing output file"), output_filename);

exit:
  fh_done ();
  i18n_done ();

  return 0;
}

static void
decrypt_sav_file (struct encrypted_sys_file *enc,
                  const char *input_filename,
                  const char *output_filename,
                  const char *password)
{
  FILE *out;
  int err;

  if (password == NULL)
    {
      password = getpass ("password: ");
      if (password == NULL)
        exit (1);
    }

  if (!encrypted_sys_file_unlock (enc, password))
    error (1, 0, _("sorry, wrong password"));

  out = fn_open (output_filename, "wb");
  if (out == NULL)
    error (1, errno, ("%s: error opening output file"), output_filename);

  for (;;)
    {
      uint8_t buffer[1024];
      size_t n;

      n = encrypted_sys_file_read (enc, buffer, sizeof buffer);
      if (n == 0)
        break;

      if (fwrite (buffer, 1, n, out) != n)
        error (1, errno, ("%s: write error"), output_filename);
    }

  err = encrypted_sys_file_close (enc);
  if (err)
    error (1, err, ("%s: read error"), input_filename);

  if (fflush (out) == EOF)
    error (1, errno, ("%s: write error"), output_filename);
  fn_close (output_filename, out);
}

static void
usage (void)
{
  printf ("\
%s, a utility for converting SPSS data files to other formats.\n\
Usage: %s [OPTION]... INPUT OUTPUT\n\
where INPUT is an SPSS system or portable file\n\
  and OUTPUT is the name of the desired output file.\n\
\n\
The desired format of OUTPUT is by default inferred from its extension:\n\
  csv txt             comma-separated value\n\
  sav sys             SPSS system file\n\
  por                 SPSS portable file\n\
\n\
Options:\n\
  -O, --output-format=FORMAT  set specific output format, where FORMAT\n\
                      is one of the extensions listed above\n\
  -e, --encoding=CHARSET  override encoding of input data file\n\
  -c MAXCASES         limit number of cases to copy (default is all cases)\n\
  -p PASSWORD         password for encrypted .sav files\n\
  --help              display this help and exit\n\
  --version           output version information and exit\n",
          program_name, program_name);
}
